### PR TITLE
[CALCITE-4269] Improvement on enumerable implementation for HOP and SESSION

### PR DESCRIPTION
**Pull request by amaliujia**
Some improvements:
1. Use ArrayDeque to replace LinkedList for the sake of efficiency.
2. Mark the list as "final".

### CODE REVIEW

**Comment by amaliujia** [comment]
cc @vlsi

**Comment by vlsi** [comment]
LGTM, thanks

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 1]
```diff
@@ -856,7 +856,7 @@ static Expression tumblingWindowSelector(
     private final int indexOfWatermarkedColumn;
     private final int indexOfKeyColumn;
     private final long gap;
-    private LinkedList<Object[]> list;
+    private final ArrayDeque<Object[]> list;
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:859]: `Deque` might be slightly better here

File: core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java [hunk 2]
```diff
@@ -998,7 +998,7 @@ private boolean isOverlapped(Pair<Long, Long> a, Pair<Long, Long> b) {
     private final long emitFrequency;
     private final long windowSize;
     private final long offset;
-    private LinkedList<Object[]> list;
+    private final ArrayDeque<Object[]> list;
```
Comment (vlsi) [inline core/src/main/java/org/apache/calcite/adapter/enumerable/EnumUtils.java:1001]: Deque?

**Pull request merged**